sycl/doc/design/OffloadDesign.md

```diff
@@ -0,0 +1,326 @@
+# Implementation design for offloading model
```

Comment:
Adding a global comment here, so the conversation can be threaded.

It seems like there is a lot of overlap with this design and the change we plan to do for AOT for the [optional kernel features](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_discard_queue_events.asciidoc).  I think this design document should at least describe how those AOT changes will be reflected in this new offload design.  We could still implement them as separate phases if we want, though I suspect it will be easier to do some of the work together.

In particular, I think we should consider adding the following "preparatory work" to this design, which will make the optional kernel features support easier:

* Fully support the interface where specific device names are provided via the `-fsycl-targets` command line option.  Currently, we allow only a single device name like `-fsycl-targets=intel_gpu_pvc`.  We should expand this to allow multiple device names such as `-fsycl-targets=intel_gpu_pvc,intel_gpu_dg1`.

* Currently, we invoke `ocloc` once for all GPU targets.  This should be changed to invoke it separately for each GPU device on the `-fsycl-targets` command line option.

* Currently, the output from `ocloc` (i.e. native code for all GPU targets) is bundled as one offload region whose type is `PI_DEVICE_BINARY_TYPE_NATIVE`.  Now that we invoke `ocloc` separately for each GPU device, will will have a separate offload region for each GPU device named on the `-fsycl-targets` command line option.  We need some more device-specific label for these offload regions other than `PI_DEVICE_BINARY_TYPE_NATIVE`.

* The SYCL runtime needs to be adjusted to find the appropriate offload region for the target device.

I think this will affect the offload design in several ways, but one that comes to mind is the `--gen-tool-arg` option that you propose.  This is probably not granular enough because `clang-linker-wrapper` will invoke `ocloc` several times, once for each GPU device.  Do we need some way to specify different arguments for each invocation?

Reply:
Good points - current proposal is very singular in nature.  In order to support the multiple targets, this will need to be broken down into meaningful pieces where we can specify an arbitrary number of different devices and given each device, a separate `ocloc` call is performed.  This should also provide a way of passing different additional options to each individual `ocloc` call, effectively enabling support for items like `-Xsycl-target-backend=intel_gpu_pvc <opts>`